pkg/kubelet/volume/reconciler/reconciler.go

```diff
@@ -0,0 +1,231 @@
+		actualStateOfWorld:            actualStateOfWorld,
+		operationExecutor:             operationExecutor,
+	}
+}
+
+type reconciler struct {
+	controllerAttachDetachEnabled bool
+	loopSleepDuration             time.Duration
+	waitForAttachTimeout          time.Duration
+	hostName                      string
+	desiredStateOfWorld           cache.DesiredStateOfWorld
+	actualStateOfWorld            cache.ActualStateOfWorld
+	operationExecutor             operationexecutor.OperationExecutor
+}
+
+func (rc *reconciler) Run(stopCh <-chan struct{}) {
+	wait.Until(rc.reconciliationLoopFunc(), rc.loopSleepDuration, stopCh)
+}
+
+func (rc *reconciler) reconciliationLoopFunc() func() {
+	return func() {
+		// Unmounts are triggered before mounts so that a volume that was
+		// referenced by a pod that was deleted and is now referenced by another
+		// pod is unmounted from the first pod before being mounted to the new
+		// pod.
+
+		// Ensure volumes that should be unmounted are unmounted.
+		for _, mountedVolume := range rc.actualStateOfWorld.GetMountedVolumes() {
+			if !rc.desiredStateOfWorld.PodExistsInVolume(mountedVolume.PodName, mountedVolume.VolumeName) {
+				// Volume is mounted, unmount it
+				glog.V(12).Infof("Attempting to start UnmountVolume for volume %q (spec.Name: %q) from pod %q (UID: %q).",
+					mountedVolume.VolumeName,
+					mountedVolume.OuterVolumeSpecName,
+					mountedVolume.PodName,
+					mountedVolume.PodUID)
+				err := rc.operationExecutor.UnmountVolume(
+					mountedVolume.MountedVolume, rc.actualStateOfWorld)
+				if err == nil {
+					glog.Infof("UnmountVolume operation started for volume %q (spec.Name: %q) from pod %q (UID: %q).",
+						mountedVolume.VolumeName,
+						mountedVolume.OuterVolumeSpecName,
+						mountedVolume.PodName,
+						mountedVolume.PodUID)
+				}
+			}
+		}
+
+		// Ensure volumes that should be attached/mounted are attached/mounted.
+		for _, volumeToMount := range rc.desiredStateOfWorld.GetVolumesToMount() {
+			volMounted, err := rc.actualStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName)
+			if cache.IsVolumeNotAttachedError(err) {
+				// Volume is not attached, it should be
+				if rc.controllerAttachDetachEnabled || !volumeToMount.PluginIsAttachable {
+					// Kubelet not responsible for attaching or this volume has a non-attachable volume plugin,
+					// so just add it to actualStateOfWorld without attach.
+					addVolumeNodeErr := rc.actualStateOfWorld.MarkVolumeAsAttached(
+						volumeToMount.VolumeSpec, rc.hostName)
+					if addVolumeNodeErr != nil {
+						glog.Errorf(
+							"actualStateOfWorld.MarkVolumeAsAttached failed for volume %q (spec.Name: %q) pod %q (UID: %q) controllerAttachDetachEnabled: %v with err: %v",
+							volumeToMount.VolumeName,
+							volumeToMount.VolumeSpec.Name(),
+							volumeToMount.PodName,
+							volumeToMount.Pod.UID,
+							rc.controllerAttachDetachEnabled,
+							addVolumeNodeErr)
+					}
```

Comment:
add an else for this case and log that you marked it as attach (at a high V level)